WalletWasabi/BitcoinCore/Mempool/MempoolMirror.cs

```diff
@@ -66,94 +56,65 @@ private void MempoolService_TransactionReceived(object? sender, SmartTransaction
 		AddTransactions(stx.Transaction);
 	}
 
-	internal int AddTransactions(params Transaction[] txs) => AddTransactions(txs as IEnumerable<Transaction>);
-
-	private int AddTransactions(IEnumerable<Transaction> txs)
+	internal int AddTransactions(params Transaction[] txs)
 	{
-		var added = 0;
 		lock (MempoolLock)
 		{
-			// Loop through the newly received transactions expect those that are already present in our mempool snapshot.
-			foreach (Transaction tx in txs.Where(x => !Mempool.ContainsKey(x.GetHash())).ToArray())
-			{
-				// Evict double spends.
-				foreach (TxIn txInput in tx.Inputs)
-				{
-					// tx spends a UTXO which some *other* transaction in the mempool *already* spends.
-					if (PrevOutsIndex.TryGetValue(txInput.PrevOut, out Transaction? mempoolTx))
-					{
-						// Remove the old transaction from our mempool snapshot.
-						RemoveTransactionNoLock(mempoolTx.GetHash());
-					}
-				}
-
-				AddTransactionNoLock(tx);
-				added++;
-			}
+			return Mempool.AddMissingTransactions(txs);
 		}
-		return added;
 	}
 
-	private void AddTransactionNoLock(Transaction tx)
+	private async Task<int> MirrorMempoolAsync(CancellationToken cancel)
 	{
-		Mempool.Add(tx.GetHash(), tx);
+		// Get all TXIDs in the up-to-date mempool.
+		uint256[]? newTxids = await Rpc.GetRawMempoolAsync(cancel).ConfigureAwait(false);
+
+		Mempool newMempool;
 
-		foreach (TxIn txInput in tx.Inputs)
+		lock (MempoolLock)
 		{
-			PrevOutsIndex.Add(txInput.PrevOut, tx);
+			newMempool = Mempool.Clone();
 		}
-	}
 
-	private void RemoveTransactionNoLock(uint256 txHash)
-	{
-		if (Mempool.Remove(txHash, out Transaction? transaction))
+		ISet<uint256> oldTxids = newMempool.GetMempoolTxids();
+
+		// Those TXIDs that are in the new mempool snapshot but not in the old one, are the ones
+		// for which we want to download the corresponding transactions via RPC.
+		IEnumerable<uint256> missingTxids = newTxids.Except(oldTxids);
+
+		// Remove those transactions that are not present in the new mempool snapshot.
+		foreach (uint256 txid in oldTxids.Except(newTxids).ToHashSet())
 		{
-			// Remove all UTXOs of the removed transaction from our helper index.
-			foreach (TxIn txInput in transaction.Inputs)
+			if (!newMempool.TryRemoveTransaction(txid))
 			{
-				_ = PrevOutsIndex.Remove(txInput.PrevOut);
+				throw new InvalidOperationException("Should not happen.");
```

Comment:
Abort the whole thing doesn't look good, in fact if the idea if to throw exception in this way the the `TryRemoveTransaction`  should be renamed as `RemoveTransaction` and throw the exception by itself. 

So, if the idea is to throw exceptions then collect all the failing cases and throw one at the end. Also, the "This is impossible"/"Should not happen" without providing anything useful is a source of frustration.